File: problem_statement_2/main.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

int main() {

    char encoded[] = "ngr%htqqfgtwfyj%nhy";
    char message[strlen(encoded) + 1];
    int len = strlen(encoded);
    message[0] = '\0';

    for (int i = 0; i < len; i++) {

        // decode character
        char decoded = encoded[i] - 5;
        message[i] = decoded;
        // check if alphabet
        if (isalpha(decoded)) {

            decoded = toupper(decoded);

            if (decoded >= 'A' && decoded <= 'F') {
                printf("UP ");
            }
            else if (decoded >= 'G' && decoded <= 'L') {
                printf("DOWN ");
            }
            else if (decoded >= 'M' && decoded <= 'R') {
                printf("LEFT ");
            }
            else if (decoded >= 'S' && decoded <= 'Z') {
                printf("RIGHT ");
            }
        }
    }

    printf("\n%s\n", message);

    return 0;
}

File: problem_statement_2/main.py
encoded = "ngr%htqqfgtwfyj%nhy"
message = ""
result = []

for ch in encoded:
    # decode character
    decoded = chr(ord(ch) - 5)

    # check if alphabet
    if decoded.isalpha():
        decoded = decoded.upper()

        if 'A' <= decoded <= 'F':
            result.append("UP")
        elif 'G' <= decoded <= 'L':
            result.append("DOWN")
        elif 'M' <= decoded <= 'R':
            result.append("LEFT")
        elif 'S' <= decoded <= 'Z':
            result.append("RIGHT")
    
    message += decoded

print(message)
print(" ".join(result))